lightclient/src/platform/wasm_socket.rs

```diff
@@ -221,16 +225,24 @@ impl AsyncWrite for WasmSocket {
         Poll::Ready(Ok(()))
     }
 
-    fn poll_close(self: Pin<&mut Self>, _: &mut Context<'_>) -> Poll<Result<(), io::Error>> {
-        Poll::Ready(Ok(()))
+    fn poll_close(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), io::Error>> {
+        if self.socket.ready_state() == web_sys::WebSocket::CLOSED {
+            return Poll::Ready(Ok(()));
+        }
+
+        if self.socket.ready_state() != web_sys::WebSocket::CLOSING {
+            let _ = self.socket.close();
+        }
+
+        let mut inner = self.inner.lock().expect("Mutex is poised; qed");
+        inner.waker = Some(cx.waker().clone());
+        Poll::Pending
     }
 }
 
 impl Drop for WasmSocket {
     fn drop(&mut self) {
-        let inner = self.inner.lock().expect("Mutex is poised; qed");
-
-        if inner.state == ConnectionState::Opened {
+        if self.socket.ready_state() == web_sys::WebSocket::OPEN {
```

Comment:
Would it be safer to say `if ready_state != CLOSING then close()`? Or maybe just call `close()` regardless to ensure our `onclose` handler always fires and can tidy up (I think nothing happens if you try closing when it's already closing anyway but not sure!)

Reply:
I've extended this to do clean-up for `OPEN` and `CONNECTING`, since we might generate console errors if we attempt to close a socket that's already closed 👍

Reply:
Attempting to close the socket regardless of its state will generate on `--chrome`:

```bash
WebSocket connection to 'ws://polkadot-boot.dwellir.com:30334/' failed: WebSocket is closed before the connection is established.
```

The same error is generated if the socket is in the `CONNECTING` state, note that `--firefox` is not printing this warning.